File: programmers/algorithm_data_structure/binary_search/입국심사.py
def my_solution(n, times):
    '''
    나올 최솟값을 목표로 해서 이분탐색한다.
    아래 경우의 수를 따져 약간 변형한다.

    times = [4, 7, 10] 이라고 하자.
    아래 시뮬레이션 돌리는 걸로 돌려보면

    1.
    최솟값(middle)이 20일 때
    n은 8, 9 둘 다 가능하지만 아래처럼 middle//time 하면 9가 나온다.
    하지만 원하는 n은 8이라면 어떻게 찾지?
    => 이 경우 그냥 목표 n이 8, 9라고 생각해도 되지 않나?
    만약 n = 8/9일 때 두 최솟값이 다르다면 예정대로 8에 갈 것이다.
    그러니까 사실 신경쓰지 않아도 되는 경우의 수

    2. 
    middle이 15일 때 result는 6이 나온다.
    하지만 15는 존재하지 않는 경우의 수다. n = 6일 때 최솟값은 14다.
    (같은 n 내에서 존재하는 최솟값은 1개이며 항상 범위 내 최소이다)
    n이 같을 때 아래 방향으로 이동하며 answer를 업데이트 하면 된다.
    '''
    start = 0
    end = max(times) * n
    answer = 0

    while start <= end:
        middle = (start + end) // 2
        result = sum([middle//time for time in times])
        if result < n:
            start = middle + 1
        elif result >= n:
            end = middle - 1
            answer = middle

    return answer


import heapq as hq


def wrong_solution(n, times):
    '''
    쌓여가는 패턴을 생각하지 않고
    시뮬레이션을 돌리는 방식

    당연히 시간 효율성이 낮다.
    문제에선 최솟값만 찾으면 되는데
    이 방법은 각각 어디로 들어가야하는지를 파악하므로
    불필요한 정보도 구한다.
    '''
    # [선택할 경우 증가 후 시간: heap 비교용, 한 번 걸리는 시간, 현재 시간]
    check = [[time, time] for idx, time in enumerate(times)]
    hq.heapify(check)
    answer = 0
    for _ in range(n):
        next_total, time = hq.heappop(check)        
        answer = next_total
        hq.heappush(check, [next_total+time,time])

    return answer

File: programmers/algorithm_data_structure/binary_search/입국심사.java
package algorithm_data_structure.binary_search;

public class 입국심사 {
    /**
     * @param n : 심사를 기다리는 사람 수
     * @param times : 각 i번째 심사위원이 한 사람을 처리하는 시간
     * @return : 총 심사시간의 최솟값
     * 
     * 최솟값을 구하기 위해 매번 사람이 (대기를 해서라도) 들어갈 장소는
     * (각 심사위원이 하던 사람 마저 처리 시간 + 이번 사람 처리하는 시간)을
     * 각 심사위원마다 따지고 최솟값을 가진 심사위원을 택하면 된다.
     * 
     * 따라서 매번 심사위원마다 위의 값을 구할려면
     * (하던 사람 포함 처리한 총 사람 수) * 처리 단위 시간 + 처리 단위 시간
     * 을 구하면 된다.
     * 
     * 여기서 매번 처리인원 수 카운트하며 선택하면 시간 초과가 발생...
     * 
     * 대신 이분탐색을 활용하여 풀이한다.
     * 목표 반환값을 k라고 하자. 
     * 심사위원마다 심사시간이 K보다 같거나 작으며,
     * 모든 심사위원이 한 번씩 더 심사할 경우 모두 K가 넘는다.
     * 따라서 K / (처리 단위 시간)으로 심사위원별로 처리 인원 수를 구하여
     * 총 처리 인원 수를 구할 수 있다.
     * 
     * 그럼 K 값의 범위를 탐색해가며 input n과 맞을 때까지 탐색하면 된다.
     * 
     * ++ 테스트 케이스 추가
     * solution(48, new int[]{1, 1, 1, 1, 1})
     * 위 테스트케이스에서는 계산 과정에서 k와 n이 일치하지 않는다.
     * 그렇지만 k > n 일 때 k 최솟값이 정답이 되므로,
     * 
     * k >= n 이 되는 결과값(총 처리시간)의 최솟값을 찾으면 된다.
     */
    public long solution(int n, int[] times) {
        long start = 0L;        
        long end = 1000000000L  * n;

        while (start <= end) {
            long middle = (start + end) / 2L;
            long peopleCount = 0L;

            for (int time : times) {
                peopleCount += middle / time;
            }
            
            if (peopleCount >= n) {
                end = middle - 1;
            } else {
                start = middle + 1;
            }
        }

        return start;
    }
}